File: use-cases/easy-ship/code/java/src/main/java/lambda/ProcessNotificationHandler.java
package lambda;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.LambdaLogger;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.amazonaws.services.lambda.runtime.events.SQSEvent;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.gson.Gson;
import lambda.utils.ApiCredentials;
import lambda.utils.OrderChangeNotification;
import lambda.utils.SPAPINotification;
import lambda.utils.StateMachineInput;
import software.amazon.awssdk.services.sfn.SfnClient;
import software.amazon.awssdk.services.sfn.model.StartExecutionRequest;
import software.amazon.awssdk.services.sfn.model.StartExecutionResponse;

import java.io.IOException;
import java.util.UUID;

import static lambda.utils.Constants.*;
import static lambda.utils.Constants.REFRESH_TOKEN_KEY_NAME;

public class ProcessNotificationHandler implements RequestHandler<SQSEvent, String> {

    public String handleRequest(SQSEvent input, Context context) {
        LambdaLogger logger = context.getLogger();
        logger.log("ProcessNotification Lambda input: " + new Gson().toJson(input));

        //Iterate over SQS messages
        //Start a Step Functions workflow execution for Order Change notifications
        for (SQSEvent.SQSMessage message : input.getRecords()) {
            logger.log(String.format("Notification body: %s", message.getBody()));

            try {
                SPAPINotification notification = mapNotification(message.getBody());
                logger.log(String.format("Notification body: %s", notification.getPayload().getOrderChangeNotification().getSummary().toString()));

                // Only process the notification if it is of type 'ORDER_CHANGE'
                if (!NOTIFICATION_TYPE_ORDER_CHANGE.equals(notification.getNotificationType())) {
                    logger.log(String.format("Notification type %s skipped", notification.getNotificationType()));
                    continue;
                }

                OrderChangeNotification orderChangeNotification = notification.getPayload().getOrderChangeNotification();

                // Only process the notification is OrderLevel
                if (!orderChangeNotification.getNotificationLevel().equals(NOTIFICATION_LEVEL_ORDER_LEVEL)) {
                    logger.log(String.format("Notification level %s skipped", orderChangeNotification.getNotificationLevel()));
                    continue;
                }

                // Skip if order status is 'Pending'
                if (orderChangeNotification.getSummary().getOrderStatus().equals(NOTIFICATION_IGNORE_ORDER_STATUS)) {
                    logger.log("This event is skipped due to Pending order status");
                    continue;
                }

                // TODO: We might need to call orderAPI to confirm if the order is EasyShip or not
                String executionArn = startStepFunctionsExecution(orderChangeNotification);
                logger.log(String.format("State machine successfully started. Execution arn: %s", executionArn));

            } catch (Exception e) {
                throw new InternalError("ProcessNotification Lambda failed", e);
            }
        }

        return "Finished processing incoming notifications";
    }

    private SPAPINotification mapNotification(String notificationBody) throws IOException {
        Gson gson = new Gson();
        return gson.fromJson(notificationBody, SPAPINotification.class);
    }

    private String startStepFunctionsExecution(OrderChangeNotification orderChangeNotification) throws JsonProcessingException {
        ObjectMapper mapper = new ObjectMapper();
        StateMachineInput input = getStateMachineInput(orderChangeNotification);
        String inputStr = mapper.writeValueAsString(input);

        StartExecutionRequest request = StartExecutionRequest.builder()
                .stateMachineArn(System.getenv(STATE_MACHINE_ARN_ENV_VARIABLE))
                .name(String.format("%s-%s",
                        orderChangeNotification.getAmazonOrderId(),
                        UUID.randomUUID()))
                .input(inputStr)
                .build();

        SfnClient stepFunctions = SfnClient.builder().build();
        StartExecutionResponse result = stepFunctions.startExecution(request);

        return result.executionArn();
    }

    private StateMachineInput getStateMachineInput(OrderChangeNotification orderChangeNotification) {
        String regionCode = System.getenv(REGION_CODE_KEY_NAME);
        String refreshToken = System.getenv(REFRESH_TOKEN_KEY_NAME);
        return  StateMachineInput.builder()
                .apiCredentials(ApiCredentials.builder()
                        .regionCode(regionCode)
                        .refreshToken(refreshToken)
                        .build())
                .amazonOrderId(orderChangeNotification.getAmazonOrderId())
                .marketplaceId(orderChangeNotification.getSummary().getMarketplaceID())
                .build();
    }
}

File: use-cases/easy-ship/code/php/lambda/ProcessNotificationHandler.php
<?php

namespace Lambda;

use Aws\Exception\AwsException;
use Aws\Sfn\SfnClient;
use Lambda\Utils\Constants;
use Lambda\Utils\LambdaContext;
use Lambda\Utils\Model\ApiCredentials;
use Lambda\Utils\Model\NotificationPayload;
use Lambda\Utils\Model\OrderChangeNotification;
use Lambda\Utils\Model\OrderChangeSummary;
use Lambda\Utils\Model\SPAPINotification;
use Lambda\Utils\Model\StateMachineInput;

class ProcessNotificationHandler
{
    public function handleRequest(array $event, LambdaContext $context): string
    {
        $logger = $context->getLogger();
        $logger->info('ProcessNotification Lambda input: ' . json_encode($event));

        // Iterate over SQS messages
        foreach ($event['Records'] as $message) {
            $logger->info(sprintf('Notification body: %s', $message['body']));

            try {
                // Map the notification
                $notification = $this->mapNotification($message['body']);

                // Only process notifications of type 'ORDER_CHANGE'
                if ($notification->getNotificationType() !== Constants::NOTIFICATION_TYPE_ORDER_CHANGE) {
                    $logger->info(sprintf('Notification type %s skipped', $notification->getNotificationType()));
                    continue;
                }

                // Skip if order status is 'Pending'
                if ($notification->getPayload()->getOrderChangeNotification()->getSummary()->getOrderStatus()
                    == Constants::NOTIFICATION_IGNORE_ORDER_STATUS) {
                    $logger->info('This event is skipped due to Pending order status');
                    continue;
                }

                $orderChangeNotification = $notification->getPayload()->getOrderChangeNotification();

                // Only process notifications with level 'ORDER_LEVEL'
                if ($orderChangeNotification->getNotificationLevel() !== Constants::NOTIFICATION_LEVEL_ORDER_LEVEL) {
                    $logger->info(sprintf(
                        'Notification level %s skipped',
                        $orderChangeNotification->getNotificationLevel()
                    ));
                    continue;
                }

                // Start Step Functions Execution
                $executionArn = $this->startStepFunctionsExecution($orderChangeNotification);
                $logger->info(sprintf('State machine successfully started. Execution ARN: %s', $executionArn));
            } catch (\Exception $e) {
                $logger->error('Error processing notification: ' . $e->getMessage());
                throw new \RuntimeException('ProcessNotification Lambda failed', 0, $e);
            }
        }

        return 'Finished processing incoming notifications';
    }

    /**
     * Maps the JSON notification body to an SPAPINotification object.
     *
     * @param string $notificationBody The JSON string representing the notification.
     *
     * @return SPAPINotification The mapped notification object.
     *
     * @throws \JsonException If JSON decoding fails.
     * @throws \Exception
     */
    private function mapNotification(string $notificationBody): SPAPINotification
    {
        $data = json_decode($notificationBody, true, 512, JSON_THROW_ON_ERROR);

        $orderChangeNotification = null;
        if (isset($data['Payload']['OrderChangeNotification'])) {
            $notificationData = $data['Payload']['OrderChangeNotification'];
            $summary = $notificationData['Summary'];
            $orderChangeSummary = isset($summary) ? new OrderChangeSummary (
                $summary['MarketplaceId'],
                $summary['OrderStatus'],
                $summary['FulfillmentType']): null;
            $orderChangeNotification = new OrderChangeNotification(
                $notificationData['NotificationLevel'] ?? '',
                '',
                $notificationData['AmazonOrderId'] ?? '',
                null,
                $orderChangeSummary
            );
        }

        return new SPAPINotification(
            $data['NotificationType'] ?? '',
            new \DateTime($data['EventTime'] ?? 'now'),
            new NotificationPayload($orderChangeNotification)
        );
    }


    /**
     * Starts an AWS Step Functions execution with the given OrderChangeNotification.
     *
     * @param OrderChangeNotification $orderChangeNotification The order change notification data.
     *
     * @return string The execution ARN of the started Step Functions execution.
     *
     * @throws \JsonException If JSON encoding fails.
     * @throws \RuntimeException If the execution fails to start.
     * @throws \Exception If an unexpected error occurs.
     */
    private function startStepFunctionsExecution(OrderChangeNotification $orderChangeNotification): string
    {
        $input = $this->getStateMachineInput($orderChangeNotification);
        $inputStr = json_encode($input, JSON_THROW_ON_ERROR);

        $sfnClient = new SfnClient([
            'region' => getenv('AWS_REGION'),
            'version' => 'latest',
        ]);

        try {
            $result = $sfnClient->startExecution([
                'stateMachineArn' => getenv(Constants::STATE_MACHINE_ARN_ENV_VARIABLE),
                'name' => sprintf('%s-%s', $orderChangeNotification->getAmazonOrderId(), uniqid()),
                'input' => $inputStr,
            ]);
            return $result['executionArn'];
        } catch (AwsException $e) {
            throw new \RuntimeException('Failed to start Step Functions execution: ' . $e->getAwsErrorMessage(), 0, $e);
        } catch (\Exception $e) {
            throw new \RuntimeException('An unexpected error occurred while starting Step Functions execution.', 0, $e);
        }
    }


    /**
     * @throws \Exception
     */
    private function getStateMachineInput(OrderChangeNotification $orderChangeNotification): StateMachineInput
    {
        $regionCode = getenv(Constants::REGION_CODE_KEY_NAME);
        $refreshToken = getenv(Constants::REFRESH_TOKEN_KEY_NAME);

        return new StateMachineInput(
            new ApiCredentials($refreshToken, $regionCode),
            $orderChangeNotification->getAmazonOrderId(),
            $orderChangeNotification->getSummary()->getMarketplaceId()
        );
    }
}
